pkg/security/podsecuritypolicy/util/util.go

```diff
@@ -242,3 +242,27 @@ func EqualStringSlices(a, b []string) bool {
 	}
 	return true
 }
+
+func IsOnlyServiceAccountTokenSources(v *api.ProjectedVolumeSource) bool {
+	for _, s := range v.Sources {
+		if t := s.ServiceAccountToken; t != nil && (t.Path != "token" || t.Audience != "") {
+			return false
+		}
+
+		if s.ConfigMap != nil && s.ConfigMap.LocalObjectReference.Name != "kube-root-ca.crt" {
+			return false
+		}
+
+		if s.DownwardAPI != nil {
+			for _, d := range s.DownwardAPI.Items {
+				if d.Path != "namespace" || d.FieldRef == nil || d.FieldRef.APIVersion != "v1" || d.FieldRef.FieldPath != "metadata.namespace" {
+					return false
+				}
+			}
+		}
+		if s.Secret != nil {
+			return false
+		}
+	}
```

Comment:
one last bit, to be future-proof against additional projected sources in the future, return false if this is not a configmap/serviceaccounttoken/downwardAPI source:

```go
// reject any projected source that does not match any of our expected source types
if s.ServiceAccountToken == nil && s.ConfigMap == nil && s.DownwardAPI == nil {
  return false
}
```

Reply:
done

Reply:
wait, this is not future proof:

this volume source will be allowed.
```
Secret: &api.SecretProjection{},
  ConfigMap: &api.ConfigMapProjection{
    LocalObjectReference: api.LocalObjectReference{
      Name: "kube-root-ca.crt",
    },
    Items: []api.KeyToPath{
      {
        Key:  "ca.crt",
        Path: "ca.crt",
       },
     },
 },
},
```

Reply:
> this volume source will be allowed.

it should be allowed... it only contains data that is exposed via an injected service account token volume

Reply:
the `if s.Secret != nil { return false }` covers that case

Reply:
right. how is this future-proof if we need to add every new projected source here along with secret?

Reply:
API validation disallows more than one projected source in a single item:

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L1084-L1086

Reply:
ah. then `if s.Secret != nil { return false` is not needed. thanks.